amd/src/container_component.js

```diff
@@ -0,0 +1,92 @@
+ *
+ * Moodle is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Moodle is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import {BaseComponent} from 'core/reactive';
+import {get_string as getString} from 'core/str';
+import Templates from 'core/templates';
+
+export default class extends BaseComponent {
+
+    /**
+     * Initializes a container component.
+     *
+     * @param {string} uniqid
+     * @returns {BaseComponent}
+     */
+    static init(uniqid) {
+        return new this({
+            element: document.getElementById(uniqid),
+        });
+    }
+
+    /**
+     * This method is called when the component gets created.
+     */
+    create() {
+        this.uniqid = this.element.dataset.uniqid;
+
+        this.selectors = {
+            HEADERALL: `#qpy-all-header-${this.uniqid}`,
+            CONTENTALL: `#qpy-all-content-${this.uniqid} > .qpy-tab-content`,
+            LOADINGINDICATOR: `#qpy-loading-indicator-${this.uniqid}`
+        };
+    }
+
+    /**
+     * This method gets called when the state is ready and is used to render the packages.
+     *
+     * @param {object} state
+     * @returns {Promise<void>}
+     */
+    async stateReady(state) {
+        const packages = JSON.parse(JSON.stringify(state.packages));
+        this._renderPackages(packages);
+    }
+
+    /**
+     * Renders package cards inside the container and hide loading indicator when finished.
+     *
+     * @param {array} packages
+     * @private
+     */
+    _renderPackages(packages) {
+        getString('all_packages', 'qtype_questionpy', packages.length).then(string => {
+            this.getElement(this.selectors.HEADERALL).innerHTML = string;
+            return null;
+        }).catch();
+
+        // Render each package.
+        for (const packageData of packages) {
+            Templates.renderForPromise('qtype_questionpy/package/package_selection', {
+                "namespace": packageData.namespace,
+                "name": packageData.name,
+                "description": packageData.description,
+                "icon": packageData.icon,
+                "selected": false,
+                "versions": packageData.versions,
+                "url": packageData.url
+            }).then(({html, js}) => {
+                Templates.appendNodeContents(this.selectors.CONTENTALL, html, js);
+                return null;
+            }).catch();
```

Comment:
Tut es das nicht auch ohne `return null` und leeres `.catch()`?

```suggestion
            Templates.renderForPromise('qtype_questionpy/package/package_selection', {
                "namespace": packageData.namespace,
                "name": packageData.name,
                "description": packageData.description,
                "icon": packageData.icon,
                "selected": false,
                "versions": packageData.versions,
                "url": packageData.url
            }).then(({html, js}) => {
                Templates.appendNodeContents(this.selectors.CONTENTALL, html, js);
            });
```